### DISCUSSION TITLE
Telemetry validation

### DISCUSSION THREAD

**Discussion by pnadolny13** [meltano/meltano#6036] [General]
Starting this discussion to talk about the telemetry changes as part of https://github.com/meltano/meltano/issues/3365 and https://github.com/meltano/meltano/issues/3366.

**Comment by pnadolny13** [comment]
@pandemicsyn @WillDaSilva I'm seeing a few things so far that I wanted to note:

1. I think theres a bug with [`is_ci_environment`](https://github.com/meltano/meltano/blob/1498bf59afa5a9eb44059a6a4f550caab4e07c31/src/meltano/core/tracking/environment.py#L58). I got `True` locally which I dont think should be the case.
2. Also for [`is_dev_build`](https://github.com/meltano/meltano/blob/1498bf59afa5a9eb44059a6a4f550caab4e07c31/src/meltano/core/tracking/environment.py#L57). I got `True` locally and I'm not 100% sure what this attribute means,  I see a file [is created in CI](https://github.com/meltano/meltano/blob/1498bf59afa5a9eb44059a6a4f550caab4e07c31/.gitlab/ci/publish.gitlab-ci.yml#L17).
3. Its probably just because were still implementing but I dont see [`subcommand` or `option_keys`](https://github.com/meltano/meltano/blob/1498bf59afa5a9eb44059a6a4f550caab4e07c31/src/meltano/core/tracking/cli.py#L32) used yet. What would I expect those to be? Is that what we were referring to as `args` in the original issue https://gitlab.com/meltano/meltano/-/issues/3443?

**Reply by pnadolny13** [reply]
Related to 3 above, if `option_keys` = `args` then I think we need to add the kwargs at https://github.com/meltano/meltano/blob/1498bf59afa5a9eb44059a6a4f550caab4e07c31/src/meltano/cli/invoke.py#L79.

**Reply by pandemicsyn** [reply]
1. Can confirm, is_ci_environment has a bug. Logged https://github.com/meltano/meltano/issues/6038
2. That should return true when using a meltano vers installed manually but false when running a pip installed release. But that may not work as intended yet (https://github.com/meltano/meltano/issues/6021) so I'd probably ignore until that resolves.
3. Yea subcommand, and option_keys aren't used on main yet. There just a quick and dirty version for invoke because @aaronsteers wanted an example of it in the merge. 

    There's a inprogress sample usage of it here: https://github.com/meltano/meltano/pull/6028/files#diff-34adbf6f8e1cd00c20d210df9990552eefc56339dae2326c11f038aed7ff2025R106

   Args is kinda overloaded and i think `option` is actually more appropriate and map's to what they're actually called (options, arguments, etc are all subtly different). So examples will be like:


```
$ meltano run --dry-run tap-gitalb target-jsonl
command = run
sub_command = None
option_keys = ["dry_run"]

$ meltano schedule set --job
command = schedule
sub_command = set
option_keys = ["job"]

$ meltano invoke something
command = invoke
sub_command = None
options_keys = []

```

**Reply by pnadolny13** [reply]
@pandemicsyn thanks thats clarifies a lot.

**Reply by pnadolny13** [reply]
@pandemicsyn @aaronsteers one other thing I noticed was that we no longer track the full command passed. Previously we were tracking command = `meltano run --environment=asdfasdf tap-x target-y` and now we have command = `run`. I think this is fine especially as we parse and send more things like  the `option_keys` and its more risky for PII to send the full command but want to make sure were intentional about that.

Would we ever want to still have that full semi-raw command? If the plugin context event isnt sent or the plugin level events fail before firing (idk if this is possible if we are firing fail events anyways) then we have no idea what was being attempted, just `run`, soon to be including the `option_keys` args but still no reference to the plugin names. Even if theyre just the plain old `tap-postgres` with no other detail.

**Reply by pandemicsyn** [reply]
That was definitely intentionally, because y'all reported that the raw command was hard to parse and could contain PII data. I assume thats why we hash even the plugin name in in the plugin context now.

**Reply by aaronsteers** [reply]
@pnadolny13 and @pnadolny13 - Yeah, that's correct. 

@pnadolny13 - To reconstruct the CLI test (approximately), you'd have to concatenate the `meltano <command> --<options[0]> --options[3], etc.`

In future, we might introduce telemetry improvements that give higher fidelity reconstruction, but for now I think this meets the requirements of visibility and keeping pii out of the flow as much as possible.

**Comment by pnadolny13** [comment]
Is [`event_uuid`](https://github.com/meltano/meltano/blob/762a2592c5da95fa6b1e91aa3c20d79725949a66/src/meltano/core/tracking/cli.py#L45) in the CLI context necessary? The Snowplow event already gets a UUID assigned to it called `event_id` at the top level. I see why we need `context_uuid` to stitch events from the same execution but I dont think we need an additional `event_id`. Is there something I might be missing?

**Reply by aaronsteers** [reply]
@pnadolny13 - I can't speak to this specific implementation, but Event UUIDs are extremely helpful because  'at least once' delivery strategies do not promise that the event will arrive _only_ once. Generally timestamp and other items can be used to dedupe but its much preferable to know explicitly that this was a unique event.

**Reply by pnadolny13** [reply]
@aaronsteers yeah I agree, we can leave it in but I'm just saying we get a uuid generated for the event already from the snowplow sdk (I think thats where its coming from). I snowplow's event_id to dedup in the warehouse with our current structured events. The truncated example below:

```
"event": {
            "app_id": null,
            "platform": "pc",
            "etl_tstamp": "2022-06-06T20:27:21.109Z",
            "collector_tstamp": "2022-06-06T20:27:21.109Z",
            "dvce_created_tstamp": "2022-06-06T20:27:21.076Z",
            "event": "unstruct",
            "event_id": "cd6bb380-c7bd-4c07-b375-0184c2ab3aa9", # THIS IS CREATED AUTOMATICALLY
            ...
            "contexts": {
                "schema": "iglu:com.snowplowanalytics.snowplow/contexts/jsonschema/1-0-0",
                "data": [
                    {
                        "schema": "iglu:com.meltano/cli_context/jsonschema/1-0-0",
                        "data": {
                            "event_uuid": "9484f45e-f9a6-49e4-affe-52076ba32490", # WE CREATE OUR OWN ALSO
                            "command": "invoke",
                            "sub_command": null,
                            "option_keys": []
                        }
                    },
```

**Reply by aaronsteers** [reply]
Gotcha. @WillDaSilva and @pandemicsyn - do you know if the built-in `event_id` is sufficient?

Either way, probably not high-pri to change at this point.

**Reply by pandemicsyn** [reply]
Looks like the built in would be sufficient. @pnadolny13 I'll drop it in https://github.com/meltano/meltano/pull/6028

**Comment by pnadolny13** [comment]
Is it useful to add `executable_hash` to the plugins_context? It might be nice to know that if we get a custom plugin with name = `foo` but it references the `dbt` executable. Idk but it crossed my mind so wanted to check.

**Reply by aaronsteers** [reply]
@pnadolny13 - We originally had `executable_path` in the collection but I asked for it to be hashed because of the PII exposure.

If we don't need it, I don't mind removing.

**Reply by pnadolny13** [reply]
@aaronsteers I was referring to the [`executable` key](https://github.com/meltano/meltano/blob/762a2592c5da95fa6b1e91aa3c20d79725949a66/src/meltano/core/bundle/discovery.yml#L3050) from the meltano.yml file. Is my example scenario not possible? Since we have everything hashed, if they use an alias for `name` like `foo-xyz` then we wont be able to unhash it because were not looking for a dbt plugin called `foo-xyz` but they cant mess with their `dbt` executable and still have it work, so we will be able to detect that its a dbt plugin because we are expecting a `dbt` executable so it can be unhashed. Does that make sense?

**Reply by aaronsteers** [reply]
My mistake. Thanks for clarifying. Yes, that would make sense, I think, to add a hashed (or unhashed) version of `plugin.executable`.

**Reply by aaronsteers** [reply]
The other way to infer this would be to look at pip_url, but that also can have varitions that make lookups difficult.

**Comment by pnadolny13** [comment]
@pandemicsyn I tested out your updates on https://github.com/meltano/meltano/pull/6028 and its all looking good. I added a few comments above to consider. Also this might be expected behavior and in progress but when I have a failure for both invoke/run I get an initialized and failed event at the "ExtractLoad" block level, which is great, but at the top level I dont get an exit event saying the command failed.

For example:

`meltano run tap-csv target-postgres` with an invalid tap-csv config. I get 3 events:
- CLI level "started"
- ExtractLoadBlocks "initialized"
- ExtractLoadBlocks "failed"
- <MISSING: CLI level "failed">

Or if I try to run a non-existant tap like `meltano invoke tap-foo` I get a started but no end/failed event.

**Reply by aaronsteers** [reply]
Spin off on absence of failure event here: https://github.com/meltano/meltano/discussions/6040

**Reply by pandemicsyn** [reply]
@pnadolny13 yea, i don't think i have any failures covered yet in run.

**Reply by pandemicsyn** [reply]
To clarify for future folks that might read this...theres no failures covered yet in run because the PR is a WIP and is being actively developed atm 😁

**Comment by pandemicsyn** [comment]
@pnadolny13 @aaronsteers fyi - I also logged a follow-up task to push cli tracker creation further up the stack https://github.com/meltano/meltano/issues/6041 . That would get us global options, potentially fire that initial cli started event automatically, and DRY's things up a bit.

**Comment by pnadolny13** [comment]
@pandemicsyn @aaronsteers testing this morning I have this experience and had a few follow up questions below.

1. `meltano run tap-csv target-jsonl` - Configured to Succeed

- CLI level "started"
- ExtractLoadBlocks "initialized"
- ExtractLoadBlocks "completed"
- CLI level "completed"
- Legacy Structured Event

2. `meltano run tap-csv target-jsonl` - tap-csv Configured to Fail

- CLI level "started"
- ExtractLoadBlocks "initialized"
- ExtractLoadBlocks "failed"
- CLI level "failed"

3. `meltano run tap-csv target-jsonl` - target-jsonl Configured to Fail. Returns `BrokenPipeError: [Errno 32] Broken pipe` because destination path doesnt exists.

- CLI level "started"
- ExtractLoadBlocks "initialized"
- <MISSING: ExtractLoadBlocks "failed">
- CLI level "failed"


4. `meltano invoke tap-csv` - Configured to Succeed

- CLI level "started"
- CLI level "completed"
- Legacy Structured Event


5. `meltano invoke tap-csv` - tap-csv Configured to Fail

- CLI level "started"
- CLI level "failed"


For scenario 2 with an invalid tap we get 4 events but scenario 3 we only get 3 events probably due to the broken pipe from the target failing. Is it possible to get that failed event? 

For scenario 4 and 5 I see the plugin context included but I was sort of expecting there to be block-esque events sent between the CLI start/complete, totaling 4 events. Like an invoker block with its own initialized/completed events for tap-csv. I couldnt find anything where we talk about it in any of the issues but I'm wondering if it will be better to do that consistently. This isnt a hard request to add that as I'm still thinking through it but since were in a time crunch I wanted to flag it early for others to consider also. cc @tayloramurphy

**Reply by pandemicsyn** [reply]
Good catch! I tested failures but all my triggered RunnerError's which got caught. Fixing up 3 is definitely do able - and I can work that into the open PR (easy peasy one line change - I bet one of y'all would have called that out during the PR review today).

For 4/5 , there's no block events there because theres no blocks being used. When you call `meltano invoke` it doesn't create and use an InvokerBlock , it loads a plugin and invokes it directly. IMHO the command is using the plugins directly so the plugin context should be attached directly to the command event. Firing two intermediate events doesn't really serve a purpose because theres no actual intermediate actor to instrument.

We could create another intermediate event type/context and fire that if y'all wanted one - but it should at least be distinct from block events and be mainly used for encapsulation. If we wanted to go that route, I might vote for doing that as a follow-on as we're sorta running out of time in the current iteration. Although I guess that sorta depends on what y'all would rather have for the 2.0 launch.

@pnadolny13 @aaronsteers @tayloramurphy  Would y'all prefer I use the rest of today to add another intermediate event and update `invoke` to fire that, or use what we currently have and instrument some of the other commands `etl`, `schedule`, etc?

**Reply by aaronsteers** [reply]
@pandemicsyn - The two things I think would be highest priority from here are:

1. Adding the rich plugin data to `meltano invoke` and/or `meltano elt`.
2. Adding coverage to other commands.

@tayloramurphy - what's your take on priority with our remaining time?

**Reply by pnadolny13** [reply]
Following up here: The outcome from our zoom call was that we're ok with this as-is because it matches how the code actually works instead of faking it to send intermediate block level events when a block is never used and were also ok with it being somewhat inconsistent right now as a tradeoff to get more coverage on other events.

- https://github.com/meltano/meltano/issues/6056 was created to track the order of command priority.
- Separately https://github.com/meltano/meltano/issues/6057 was created to add context_uuid to the structured legacy events so we could tie the full session together if needed due to unexpected bugs or lost events (i.e. structured events fire but not unstructured).

**Comment by pnadolny13** [comment]
@pandemicsyn when I try to `run` a plugin that doesnt exist using `meltano run tap-csv target-foo-xyz` I only get a CLI started event. For `invoke` like `meltano invoke target-foo-xyz` I get a CLI started and aborted event. I would expect run to also have an "aborted" or "failed" event in this case.

**Comment by pnadolny13** [comment]
@pandemicsyn @WillDaSilva I'm deep in the weeds exploring the new telemetry data and I noticed something thats tripping me up a bit and could use some context of how the code works.

For this example I'm filtering for:
- 2.1.0 version
- invoke events only

I roll up all of the unstruct events to a single row which I'm calling an "execution". I can connect executions to their equivalent struct event to compare the data for consistency. For QA purposes of both my new dbt models and of the overall telemetry data were collecting, I'd like to be able to say "I see 100 struct events and also 100 unstruct executions" but I cant find a way to do that right now. The challenge is I cant find a way to filter the unstruct executions records in a way that ties out. We get far more unstruct executions due to the fact that we collect failed/aborted executions which we dont do with struct events. This led me to try to filter out non-successful executions but they still dont tie out.

According to the exit codes it almost looks like the struct events are firing even on unsuccessful executions but inconsistently. First of all, what are the `exit_codes` telling me? Taking dbt as an example where 5 models run and complete but 1 errors because of a typo, does the exit code reflect the dbt's exit code or meltano's exit code? The meltano command finished without errors but the plugin failed?

Heres an example of what I'm seeing: the exit code is non-zero, we only see a started/failed states, and we sent 2 cli events + 1 exit event + 1 struct event. 

EXIT_CODE | EXCEPTION_TYPE | EXCEPTION_CAUSE | STRUCT_COMAND_CATEGORY | STRUCT_COMMAND | EVENT_STATES | EVENT_NAMES
-- | -- | -- | -- | -- | -- | --
1 | SystemExit |   | meltano invoke | meltano invoke dbt --models some_model | [   "failed",   "started" ] | [   "event",   "exit_event",   "cli_event" ]
1 | SystemExit |   | meltano invoke | meltano invoke dbt-postgres  --environment=ef260e9aa3c673af240d17a2660480361a8e081d1ffeca2a5ed0e3219fc18567 | [   "failed",   "started" ] | [   "cli_event",   "event",   "exit_event" ]
1 | SystemExit |   | meltano invoke | meltano invoke dbt-postgres  --environment=ef260e9aa3c673af240d17a2660480361a8e081d1ffeca2a5ed0e3219fc18567 | [   "failed",   "started" ] | [   "cli_event",   "exit_event",   "event" ]
1 | SystemExit |   | meltano invoke | meltano invoke dbt-postgres  --environment=ef260e9aa3c673af240d17a2660480361a8e081d1ffeca2a5ed0e3219fc18567 | [   "failed",   "started" ] | [   "cli_event",   "exit_event",   "event" ]
1 | SystemExit |   | meltano invoke | meltano invoke dbt-postgres  --environment=ef260e9aa3c673af240d17a2660480361a8e081d1ffeca2a5ed0e3219fc18567 | [   "started",   "failed" ] | [   "cli_event",   "event",   "exit_event" ]
1 | SystemExit |   | meltano invoke | meltano invoke dbt-postgres  --environment=ef260e9aa3c673af240d17a2660480361a8e081d1ffeca2a5ed0e3219fc18567 | [   "started",   "failed" ] | [   "cli_event",   "exit_event",   "event" ]
1 | SystemExit |   | meltano invoke | meltano invoke tap-csv --discover --environment=ef260e9aa3c673af240d17a2660480361a8e081d1ffeca2a5ed0e3219fc18567 | [   "failed",   "started" ] | [   "cli_event",   "event",   "exit_event" ]

Does that example make sense?

cc @tayloramurphy

**Reply by pnadolny13** [reply]
Right now I can use the fact that a struct event was found via the execution ID to tie out the counts but once we turn off the struct events I wont have a way to compare apples to apples between the new and old events. I know comparing against historical data is a lower priority but its probably something we'd want to be able to support before we turn off the struct events.

**Reply by pnadolny13** [reply]
It looks like we fire the struct event without checking the exit code here https://github.com/meltano/meltano/blob/c1792b420248af3a25795c1e82568b2f761db95b/src/meltano/cli/invoke.py#L195. Is that only getting skipped if an exception is thrown? When would an exception be thrown vs just a non-zero exit code?

Its possible this is a bug with the struct events, which we probably wouldnt want to fix but its the first time we have enough information to notice it.

**Reply by pandemicsyn** [reply]
> According to the exit codes it almost looks like the struct events are firing even on unsuccessful executions but inconsistently. F

yes because in invoke firing the "old" structured event was not tied to an exit code. Whether or not it fired was entirely dependent on what sort of error happened and whether it was caught up stream or invoke actually ran long enough to fire a GA event. Most of the legacy events were what i'd call "best-effort" in that regard. 

>  Is that only getting skipped if an exception is thrown? When would an exception be thrown vs just a non-zero exit code?

Yep. If you invoke tap-gitlab and it exits with a exit code of "1" we use that as our exit code and both unstruct and struct events would fire. However, if during the course of trying to invoke something we an encounter an error (e.g. UnknownCommandError, a broken pipe etc) the legacy structured event would never get fired.

**Reply by pandemicsyn** [reply]
> I'd like to be able to say "I see 100 struct events and also 100 unstruct executions"

That can't work because the legacy struct events have never been reliably emitted in all scenarios.  Wasn't the fact that they often only fired for successful meltano executions one of the big drivers for all this new work?

 IMHO the best you could do is make sure that you're seeing **at least** **as many** unstruct events as legacy struct events. If you run into a case where you have consistently less, it means we have a gap in coverage for the new tracking.

**Comment by pandemicsyn** [comment]
I think @pnadolny13 might have actually reported this as well above, but I think there are some issues with the exit events and exception tracking. I'm experiencing some confusing behavior. 

Raw event dump for a single invocation of `meltano run --dry-run tap-gitlab target-jsonl` (this issue isn't limited to `meltano run`): https://gist.github.com/pandemicsyn/6fbc5cd0d045baf7f8605d3e4ea70a89 

1. There's a `SystemExit` exception recorded even on _successful_ invocations of the cli ([link to relevant line number of the gist](https://gist.github.com/pandemicsyn/6fbc5cd0d045baf7f8605d3e4ea70a89#file-exit-events-somuch-json-L167)). @pnadolny13 I think these are "benign", but IMHO we shouldn't even be sending those to you - since you'd basically just need to filter these out anyway based on the contents of the fields or other contexts. 

   ```json
                    {
                        "schema": "iglu:com.meltano/exception_context/jsonschema/1-0-0",
                        "data": {
                            "context_uuid": "3ffa5280-6ba1-4c51-9589-a9cdc07866c8",
                            "exception": {
                                "type": "SystemExit",
                                "str_hash": "5feceb66ffc86f38d952786c6d696c79c2dbc239dd4e91b46729d73a27fb57e9",
                                "repr_hash": "add974c1324d8837f81b9347ec532f5b9621742ea62382fc7e56fcf03d996454",
                                "traceback": [
                                    {
                                        "file": ".../cli/__init__.py",
                                        "line_number": 85
                                    },
                                    {
                                        "file": ".../cli/__init__.py",
                                        "line_number": 66
                                    },
                                    {
                                        "file": "lib/python3.8/site-packages/click/core.py",
                                        "line_number": 1130
                                    },
                                    {
                                        "file": ".../cli.py",
                                        "line_number": 32
                                    },
                                    {
                                        "file": "lib/python3.8/site-packages/click/core.py",
                                        "line_number": 1083
                                    }
                                ],
                                "cause": null,
                                "context": {
                                    "type": "Exit",
                                    "str_hash": "5feceb66ffc86f38d952786c6d696c79c2dbc239dd4e91b46729d73a27fb57e9",
                                    "repr_hash": "3b457417f1371a2ff2a1e4f1d4dab1b67431f91c58617b23c31f1b4b2250e4d8",
                                    "traceback": [
                                        {
                                            "file": "lib/python3.8/site-packages/click/core.py",
                                            "line_number": 1065
                                        },
                                        {
                                            "file": "lib/python3.8/site-packages/click/core.py",
                                            "line_number": 687
                                        }
                                    ],
                                    "cause": null,
                                    "context": null
                                }
                            }
   ```

3. I get empty exception contexts in later events, sometimes singles, sometimes multiple - in a single event ([link to relevant line number of the gist](https://gist.github.com/pandemicsyn/6fbc5cd0d045baf7f8605d3e4ea70a89#file-exit-events-somuch-json-L477))



   ```json
                    {
                        "schema": "iglu:com.meltano/exception_context/jsonschema/1-0-0",
                        "data": {
                            "context_uuid": "17ad7960-6870-48c3-a4b7-051f44f551b1",
                            "exception": null
                        }
                    },
                    {
                        "schema": "iglu:com.meltano/exception_context/jsonschema/1-0-0",
                        "data": {
                            "context_uuid": "60ecc7fa-9215-460f-8916-41f1b831f4d5",
                            "exception": null
                        }
                    }
   ```
   Additional empty singular exception contexts: [1](https://gist.github.com/pandemicsyn/6fbc5cd0d045baf7f8605d3e4ea70a89#file-exit-events-somuch-json-L748), [2](https://gist.github.com/pandemicsyn/6fbc5cd0d045baf7f8605d3e4ea70a89#file-exit-events-somuch-json-L1049)


@WillDaSilva are these two behaviors intentional/expected or defects?


side note: Generated this with what's currently on main (e655ff6623235f0c9f7b9d6a89e68fcf199ce7a8)

**Reply by WillDaSilva** [reply]
@pandemicsyn All of this behaviour seems expected.

1. Unless Python has a hard crash (e.g. segfault) it will exit with an exception, usually `SystemExit`. Note that Python exceptions are not "errors", and should not be thought of as such. There's a saying in the Python community: "exceptions are not exceptional". They're just a control flow mechanism.
2. The exception context is special in that it's generated when/where each event is fired, and not saved for future events. We generate it on-demand to make it automatic, and to capture the current state of the process (i.e. what exceptions are currently being processed). We do not save it for future events because that could easily result in a lot of data being sent with each event in a long-running process, and because there is nothing to gain by repeatedly sending exception context objects we have already delivered. If for whatever reason you care to see all of the exception contexts that were emitted by a single process run, you can group them using the context ID from the environment context each one was fired with, or whatever other shared identifier.

**Reply by pandemicsyn** [reply]
So basically when looking at the data, ignore null exception contexts inline of events, and ignore SystemExits when the exit event recorded an exit code of 0?

**Reply by WillDaSilva** [reply]
Yup, unless you care about what was going on in the process prior to its successful exit (e.g. "what prompted this successful exit?").

**Reply by pnadolny13** [reply]
@WillDaSilva yeah I hear you that those are technically valid exception but I'm wondering what value we get out of having them? Do you have use cases in mind? I'm not against having them but if its not adding value then it would probably be preferred to not pollute the events with extra stuff. So I'm mostly curious if you guys think it would potentially add value.

For the `SystemExit` exceptions - would this help us debugging in some way that I'm not aware of? If its successful then wouldnt the traceback not be all that useful?

For the null exceptions - If its not telling us anything went wrong and no other details about the exception, other than that it happened, what insights would we be getting from this?

**Reply by WillDaSilva** [reply]
> For the `SystemExit` exceptions - would this help us debugging in some way that I'm not aware of? If its successful then wouldnt the traceback not be all that useful?

Exit codes are not all that reliable. Imagine if a plugin called `exit()` while processing a fatal error - this would result in an exit code of 0. It ought to run `exit(1)`, or just raise an exception, but regardless it's easy to see how a mistake like this could be made. In these cases having the traceback can tell us a lot.

In general I think we should err towards including more anonymized data than less, because we can not go back in time to collect data if we later decide that we ought to be collecting something. Consider #6269, which would not have been a problem in the first place had we opted for more info rather than less (albeit in that case it was over a privacy concern).

If you feel strongly about this, we can filter out these exception contexts prior to sending them. I'd rather not add this special case though.

> For the null exceptions - If its not telling us anything went wrong and no other details about the exception, other than that it happened, what insights would we be getting from this?

For the null exception contexts, if the consistency of always having an exception context is not helpful, we can easily stop sending them. Entirely up to you.

**Reply by pnadolny13** [reply]
@WillDaSilva thanks! Thats helpful. I dont feel strongly and I agree that more is better than not enough but of course we dont want to send stuff that doesnt add value in some way. It sounds like there is value.

Also for context, from my perspective of not knowing the implementation details of the telemetry code its not obvious what would be a special case where were writing one off inconsistent handling of when to fire events. Its helpful to hear the context and the trade offs based on how it would be implemented when I make telemetry requests. I definitely favor consistency and correctness with the tradeoff that we get extra data that needs to be filtered out.

**Comment by pnadolny13** [comment]
@WillDaSilva @pandemicsyn I can create a separate discussion/issues for this but I'm curious if theres already a mechanism for explicitly tracking features in our current schema, related to https://github.com/meltano/squared/issues/115. I can infer certain features are used based on activity but some are implicit so we'd need to send a flag that its used.

Do we have a way to understand if a plugin thats being executed is backed by a lock file? It would be useful to know how many people are using lock files vs the legacy discovery.yml.

I'm sure there are other features we'd want to track that need to be explicitly flagged, should this be a one off or should we add some sort of features list? I'm not totally sure if thats useful if the features are at different levels (plugin vs CLI vs block).

**Reply by WillDaSilva** [reply]
> Do we have a way to understand if a plugin thats being executed is backed by a lock file? [...]

Not yet.

We'll probably want to create a feature context for this kind of thing. Or we could extend the project context with info like this. Either way we'll need to compile a list of the features we want to track. I've subscribed to https://github.com/meltano/squared/issues/115, and suggest we convert it to a discussion, or otherwise continue the discussion over there.

It may be the case that we don't need a separate feature context in the end. We may find that each feature in our list belongs in one of the pre-existing contexts.
